### ISSUE TITLE
Using keyword arguments in pairwise2.align.globalxx

### ISSUE THREAD

**Issue by hwalinga** [biopython/biopython#2328]
### Setup

I am reporting a problem with Biopython version, Python version, and operating
system as follows:

```python
>>> import sys; print(sys.version)
3.5.3 (default, Sep 27 2018, 17:25:39)
[GCC 6.3.0 20170516]
>>> import platform; print(platform.python_implementation()); print(platform.platform())
CPython
Linux-4.19.0-0.bpo.6-amd64-x86_64-with-LinuxMint-3-cindy
>>> import Bio; print(Bio.__version__)
1.74
```

(*Please copy and run the above in your Python, and copy-and-paste the output*)

So, when using the pairwise2.align function you have to pass in values to calculate the score of the alignment. These are 4 values, and it quickly becomes unreadable what these values actually are. It would be nice if we could pass them as keyword values. So, like this:

```python
pairwise2.align.globalms("ACGT", "AGGT", match=1, mismatch=-1, open=-1, extend=-1)
```

Instead of:

```python
pairwise2.align.globalms("ACGT", "AGGT", 1, -1, -1, -1)
```

**Comment by peterjc** [comment]
This is some very old code, and nowadays would not be written in this style. However, from a quick look it probably would be possible to do what you want by editing this part of the code:

https://github.com/biopython/biopython/blob/biopython-175/Bio/pairwise2.py#L365

However, you might find the new pairwise alignment code in ``Bio.Align`` since Biopython 1.72 a suitable replacement?

**Comment by hwalinga** [comment]
I have taken a brief look at `Bio.Align.PairwiseAligner`, but it seems you need to change the scoring function by changing the attributes of a class. 

```
>>> aligner.open_gap_score = -0.5
>>> aligner.extend_gap_score = -0.1
>>> aligner.target_end_gap_score = 0.0
>>> aligner.query_end_gap_score = 0.0
```

Yes, it is more clear. However, it would still be nice to use keyword arguments instead. 

But would you say using pairwise2 is discouraged? Since, I really like the localxx, and globalxx syntax where xx can specify different kinds of alignment method. That would be a lot easier instead of figuring out what specific values you need for this massive list: 

```
end_extend_gap_score, end_gap_score, end_open_gap_score, epsilon, extend_gap_score, gap_score, internal_extend_gap_score, internal_gap_score, internal_open_gap_score, left_extend_gap_score, left_gap_score, left_open_gap_score, match, match_score, mismatch, mismatch_score, mode, open_gap_score, query_end_extend_gap_score, query_end_gap_score, query_end_open_gap_score, query_extend_gap_score, query_gap_score, query_internal_extend_gap_score, query_internal_gap_score, query_internal_open_gap_score, query_left_extend_gap_score, query_left_gap_score, query_left_open_gap_score, query_open_gap_score, query_right_extend_gap_score, query_right_gap_score, query_right_open_gap_score, right_extend_gap_score, right_gap_score, right_open_gap_score, substitution_matrix, target_end_extend_gap_score, target_end_gap_score, target_end_open_gap_score, target_extend_gap_score, target_gap_score, target_internal_extend_gap_score, target_internal_gap_score, target_internal_open_gap_score, target_left_extend_gap_score, target_left_gap_score, target_left_open_gap_score, target_open_gap_score, target_right_extend_gap_score, target_right_gap_score, target_right_open_gap_score
```

pairwise2.py is indeed very old (18 years) and is definitely oddly written, especially from an OOP stand point. Dynamically defining the behavior using `__getattr__` and `__call__`. Took some time before I understood it. However, from a user standpoint it is a very neat API.

**Comment by MarkusPiotrowski** [comment]
@hwalinga I'm taking care of `pairwise2` for a few years now and wouldn't say that the use of it is discouraged. As you said, it's a neat API and it's easy to use in script-like code (once you are familiar with the module).

If you do many alignments on long sequences you may want to use the new `PairwiseAligner`, it's not only (much) faster, also you still have the alignment object after the doing the alignment. E.g. by just looking at a `pairwise2` result, you know nothing about the used alignment parameters anymore (e.g. `format_alignment` in `pairwise2` does not know if you did a local or a global alignment).

The dynamic definition of the align methods makes it a little harder to implement keyword parameters. I will have a look at it.

**Comment by mdehoon** [comment]
@hwalinga 

Just to give some background on why the `PairwiseAligner` class in `Bio.Align` was designed the way it is:

> Since, I really like the localxx, and globalxx syntax where xx can specify different kinds of alignment method. That would be a lot easier instead of figuring out what specific values you need for this massive list:

```
end_extend_gap_score, end_gap_score, end_open_gap_score, epsilon, extend_gap_score, gap_score, internal_extend_gap_score, internal_gap_score, internal_open_gap_score, left_extend_gap_score, left_gap_score, left_open_gap_score, match, match_score, mismatch, mismatch_score, mode, open_gap_score, query_end_extend_gap_score, query_end_gap_score, query_end_open_gap_score, query_extend_gap_score, query_gap_score, query_internal_extend_gap_score, query_internal_gap_score, query_internal_open_gap_score, query_left_extend_gap_score, query_left_gap_score, query_left_open_gap_score, query_open_gap_score, query_right_extend_gap_score, query_right_gap_score, query_right_open_gap_score, right_extend_gap_score, right_gap_score, right_open_gap_score, substitution_matrix, target_end_extend_gap_score, target_end_gap_score, target_end_open_gap_score, target_extend_gap_score, target_gap_score, target_internal_extend_gap_score, target_internal_gap_score, target_internal_open_gap_score, target_left_extend_gap_score, target_left_gap_score, target_left_open_gap_score, target_open_gap_score, target_right_extend_gap_score, target_right_gap_score, target_right_open_gap_score
```

In total there are 12 numbers that define the gap scoring scheme:
 - 2 for target or query
 - 2 for gap open or gap extend
 - 3 depending on the gap being on the left, within, or on the right of the sequence being aligned
This gives 2 x 2 x 3 = 12 basic attributes. As `Bio.pairwise2` does not provide the capability to set each of these gap scores separately, the `globalxx` etc. functions in `Bio.pairwise2` take fewer arguments.

As you will notice, the number of attributes of a `PairwiseAligner` object is larger than 12. This is because we provide additional attributes that simultaneously set specific subsets of these 12 gap scores, as in practice in most cases many of these 12 gap scores will be set to the same number anyway. So for example you can do
```
>>> aligner = PairwiseAligner()
>>> aligner.target_gap_score = -2
>>> aligner.query_gap_score = -4
>>> alignments = aligner.align(seq1, seq2)
```
without having to set each of the 12 attributes separately.

So (regardless of whether you use a object-oriented approach as in `Bio.Align`, or a functional approach as in `Bio.pairwise2`), the options are
 - Keep it simple and provide access to a limited number of gap score parameters only (as in `Bio.pairwise2`);
 - Provide access to all 12 gap score parameters (requiring users to set all 12 of them);
 - Provide access to all 12 gap score parameters, and also provide meta-parameters to allow users to set and query subsets of those 12.

The `PairwiseAligner` in `Bio.Align` uses the third option as it provides full flexibility to users who need it, while being user-friendly for other users.

The details of the gap scoring scheme are explained in the Biopython tutorial. I don't know if you have looked at that already, but please let us know if you have any suggestions for improvement of the documentation.

If you think it's helpful, we could change the docstring of `PairwiseAligner` to include one or two basic examples of gap scores. Currently I think it's a bit hard to understand from the docstring how gap scores are defined in `PairwiseAligner`.

**Comment by mdehoon** [comment]
@hwalinga 

The second difference between the `Bio.pairwise2` and the `PairwiseAligner` in `Bio.Align` is that the former uses functions and the latter an object-oriented design. The reasons for this are practical:
- Speed: with a function, Python needs to figure out the gap score settings every time you want to align two sequences. This may not matter if you want to align just a few sequences, but makes a real difference if you want to align millions of sequences.
- Readability: if your problem at hand requires a complex gap scoring scheme, then a function would get a long list of arguments that clutters the code. In particular if the function call appears somewhere in a loop, it can make the code unreadable and error-prone. With the object-oriented design,  you need to set the gap scores once. It results in cleaner code as it separates setting the alignment parameters from performing the alignments.

**Comment by mdehoon** [comment]
@hwalinga 

> But would you say using `pairwise2` is discouraged?

With all due respect for @MarkusPiotrowski  who has been doing an excellent job of maintaining this code, I do think it is better to transition to the `PairwiseAligner` in `Bio.Align`. I am of course heavily biased, but anyway here are the two main problems I see in `Bio.pairwise2`:

- The design of `Bio.pairwise2` is unpythonic . Parsing function names instead of arguments makes no sense in Python; things like `pairwise2.align.globalms` look strange and are hard to explain (for example where does the `align` between `pairwise2` and `globalms` come from?).  This gives a bad impression and makes us look like we don't know how to write Python code.
- Argument values are ambiguous and therefore error-prone. To take an example from the docstring: `pairwise2.align.globalms("ACCGT", "ACG", 2, -1, -.5, -.1)`; without reading the documentation it is unclear what values the 2, -1, -.5, -.1 represent.

and if you try to fix `Bio.pairwise2` to address these points, you'll probably end up with something similar to the `PairwiseAligner` in `Bio.Align`.

**Comment by mdehoon** [comment]
@hwalinga 

To make the `PairwiseAligner` easier to use, we could consider adding functions to `Bio.Align` so you could do something along the lines of
```python
>>> from Bio import Align
>>> alignments = Align.pairwise_align("global", seq1, seq2, gap_score=-1)
```

**Comment by MarkusPiotrowski** [comment]
With the recent updates in `PairwiseAligner` so that you can also use it to align any objects (e.g. three-letter sequences), the main reason for keeping `pairwise2` that remains is backward compatibility. Ease of use is another but minor reason (@mdehoon: I like the idea of adding class methods to `PairwiseAligner`).

I mentioned in an earlier discussion that the next step for `pairwise2` is probably converting it into an 'interface' for `PairwiseAligner`. Performance of `PairwiseAligner` is much better but I think that we need to keep a module name `pairwise2` for still some time. I guess it's too early to declare it deprecated.

**Comment by hwalinga** [comment]
At first, thank you for the excellent discussion on this proposal.

---

@mdehoon

>Argument values are ambiguous and therefore error-prone. To take an example from the docstring: pairwise2.align.globalms("ACCGT", "ACG", 2, -1, -.5, -.1); without reading the documentation it is unclear what values the 2, -1, -.5, -.1 represent.

Well, that's the exact reason I raised this issue :).

---

I think that having kwargs in the PairwiseAligner.align function would on itself already be a big improvement. Also, not that hard to do. In fact, I coded it already:

https://github.com/hwalinga/biopython/tree/PairwiseAligner-add-kwargs

Good for a PR? Still needs tests and documentation though.

(By the way, the diff is a lot bigger than my changes because of the black reformat thing. These are seperate commits.)

---

PS. I also wanted the align function to work as both a method and as classmethod. This however, is a bit more complicated and required writing a custom decorator. It was an interesting challenge, but might be too confusing code to be added to a publicily maintained repo. Judge yourself:

https://github.com/hwalinga/biopython/tree/kwargs-in-PairwiseAligner

In the above branch I also implemented a way to reduce code duplication in the align and score function. It does use a contextmanager (which is yet another fancy Python feature), so I doubted if it was good idea to implement it publicily as well.

---

PS2. When looking into Bio/Align/_aligners.c I discovered there was a target_gap_function and a query_gap_function. However, setting these does not work. I don't know if this issue is already known? It is also an undocumented feature. This lacking feature would make pairwise2 have functionality currently not available in Bio.Align.

---

PS3. @MarkusPiotrowski Allthough I have only a very brief exposure to the field of bioinformatics, but I have the feeling pairwise2 is very wide spread. Despite that, I think it would still be a good idea to put some deprecation warning on it, even if you wouldn't really have the intention to actually deprecate it on the short time. Off course, that can only really be done if above issues are dealt with.

**Comment by mdehoon** [comment]
@hwalinga 

> PS2. When looking into Bio/Align/_aligners.c I discovered there was a target_gap_function and a query_gap_function. However, setting these does not work. I don't know if this issue is already known? It is also an undocumented feature. This lacking feature would make pairwise2 have functionality currently not available in Bio.Align.

The `target_gap_function` and `query_gap_function` functions are documented in section 6.5.2.5 (General gap scores) of the Biopython tutorial. Please see the example in the documentation on how to use these functions. They provide the same functionality as in `Bio.pairwise2`.

**Comment by mdehoon** [comment]
> I think that having kwargs in the PairwiseAligner.align function would on itself already be a big improvement. Also, not that hard to do. In fact, I coded it already:
> https://github.com/hwalinga/biopython/tree/PairwiseAligner-add-kwargs
> Good for a PR? Still needs tests and documentation though.

This implementation adds keyword arguments to the `PairwiseAligner` method, as in
```python
>>> from Bio import Align
>>> aligner = Align.PairwiseAligner()
>>> alignments = aligner.align("ACGT","TCGT", gap_score=-2)
```
(use gap_score = -2 for this alignment, but don't change the value of gap_score attribute of the aligner object).
I am not sure what the benefit of this is, compared to this:
```python
>>> from Bio import Align
>>> aligner = Align.PairwiseAligner()
>>> aligner.gap_score = -2
>>> alignments = aligner.align("ACGT","TCGT")
```
A better solution would be the ability to set the gap_score (or any other attribute) when you create the `PairwiseAligner`, as in
```python
>>> from Bio import Align
>>> aligner = Align.PairwiseAligner(gap_score = -2)
>>> alignments = aligner.align("ACGT","TCGT")
```

**Comment by hwalinga** [comment]
@mdehoon 

Oops, assumed setting the function was set using the query_gap_function attribute directly instead of setting a function to query_gap_score. Also, the only documentation I looked into was the API docs. My bad.

---

Anyway, it indeed makes indeed more sense put it in the constructor. Also, a lot less code needed for that. 

My train of thought was more from a stand point of readable one-liners. But that is also from the idea that it could be used as a class method. 

Anyway it works:

https://github.com/hwalinga/biopython/tree/kwargs-in-PairwiseAligner-constructor

Ready now?

**Comment by mdehoon** [comment]
@hwalinga  It looks good to me. Can you prepare a pull request that also includes updates to the documentation, and some tests?

**Comment by mdehoon** [comment]
@hwalinga 

I realized there may be a problem in this code:
```python
    def __init__(self, **kwargs):
        super().__init__()
        for name, value in kwargs.items():
            setattr(self, name, value)
```
The order in which you get the `name`, `value` pairs from the `kwargs` dictionary is undefined.
If a user does
```python
>>> aligner = PairwiseAligner(target_gap_score=-2, open_gap_score=-4)
```
then depending on whether `target_gap_score` or `open_gap_score` is handled first, the `aligner` will get a different value for `target_open_gap_score`.

**Comment by hwalinga** [comment]
@mdehoon 

That's a good point. If we make this clear in the documentation, do you think it is still a problem? 

Also, For Python 3.6/3.7* and later, the ordering of the dictionary is equal to the order it is created. So, for your example, in those Python versions, `target_gap_score` is always handled first. So, that's good to know and saves the headache, but I don't know on what tempo you are dropping versions before 3.6/3.7?

\* For Python 3.6 this is an implementation detail. For Python 3.7 and later this has become a language feature.

**Comment by peterjc** [comment]
We're dropping Python 2.7 in early 2020, but don't yet have a timeline for dropping Python 3.5 - something I was thinking of suggesting anyway.

**Comment by hwalinga** [comment]
So, what is the plan? Wait till dictionaries are ordered in Python 3.6? (Which is still only true for CPython, any other implementation that supports 3.6 doesn't have to do that since it isn't a language feature in 3.6, and I don't know how long it will take for 3.6 to be dropped as well.) Or just implement it anyway and put a mention about it in the documentation?

**Comment by peterjc** [comment]
PyPy also does sorted dictionaries. Are there any other major implementations supporting the Python 3.5 or 3.6 language we need consider?

**Comment by mdehoon** [comment]
@hwalinga 

> That's a good point. If we make this clear in the documentation, do you think it is still a problem?

Yes it is still a problem: The behavior of this function would be undefined. Since the function is a convenience feature only and not a necessity, I don't think we should add it if its behavior is unpredictable.

> Also, For Python 3.6/3.7* and later, the ordering of the dictionary is equal to the order it is created. 
OK, then we can revisit this once we drop Python 3.5.

> on what tempo you are dropping versions before 3.6/3.7?

I can only speak for myself, but personally I am in favor of dropping older versions and taking advantage of the features offered by recent versions of Python.

**Comment by peterjc** [comment]
See mailing list - I am proposing dropping Python 3.5 at the same time as Python 2.7: 
https://mailman.open-bio.org/pipermail/biopython/2019-November/016731.html

**Comment by hwalinga** [comment]
Okay, then just ping me when you are about to drop Python 3.5.

**Comment by mdehoon** [comment]
@hwalinga  We have now dropped Python 3.5.

**Comment by MarkusPiotrowski** [comment]
I think this issue has been solved by #2569 and #2603. Closing now.

**Comment by mdehoon** [comment]
@hwalinga 

We are adding some convenience functions to `Bio.Align`, which similar functionality as the functions provided by `Bio.pairwise2`. If you have any comments or suggestions for these convenience functions, please let us know at issue #5290.
